Add server option for serving only artifacts and proxied serving mode

BenWilson2 opened the pull request:
## What changes are proposed in this pull request?

Add the flags: 
--serve-artifact-opt to enable starting an MLflow server for artifact serving purposes
--artifacts-only to enable artifact serving as an exclusive option to an MLflow server instance, disabling all other endpoints in the tracking service (/api/2.0/mlflow/* disabled, /api/2.0/mlflow-artifacts/* enabled only).

This is a follow-on to [4946]( https://github.com/mlflow/mlflow/pull/4946).

## How is this patch tested?

unit tests

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
Add capability to enable or disable mlflow-artifact functionality for the mlflow server, as well as the ability to enforce exclusively functionality of artifact handling for an mlflow server instance.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [x] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

BenWilson2 committed:
Add --serve-artifacts-opt and --artifacts-only options to mlflow server

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'master' of https://github.com/mlflow/mlflow into support-serve-artifacts

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Update examples to show cli arguments in MLflow Artifacts

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Update examples to turn on the REST API endpoints

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
linting

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

harupy commented on mlflow/cli.py:
```diff
@@ -320,6 +320,25 @@ def _validate_static_prefix(ctx, param, value):  # pylint: disable=unused-argume
     " used, then this option is required.",
 )
+@click.option(
+    "--serve-artifacts-opt",
```
```suggestion
    "--serve-artifacts",
```

Can we remove `-opt`?

BenWilson2 replied:
It was just to prevent the linter warning for shadowing a function name but the -opt suffix really doesn't make much sense from an optional arg perspective. I've removed it and I'll push changes soon.

harupy commented on mlflow/server/handlers.py:
```diff
@@ -262,6 +262,43 @@ def wrapper(*args, **kwargs):
 ]
 
 
+def _disable_mlflow_artifacts_endpoint(func):
```
```suggestion
def _disable_unless_serve_artifacts(func):
# I'm not sure if this sounds fine...
```

This is not a big issue, but can we rename the function to indicate that the endpoint disablement is conditional?

BenWilson2 replied:
I like it. Refactored.

BenWilson2 committed:
PR changes and adjust test for serve-artifact option flag

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'master' of https://github.com/mlflow/mlflow into support-serve-artifacts

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

harupy commented on mlflow/server/handlers.py:
```diff
@@ -262,6 +262,43 @@ def wrapper(*args, **kwargs):
+    return wrapper
+
+
+def _disable_mlflow_artifacts_only(func):
```
```suggestion
def _disable_if_artifacts_only(func):
```

Can we also rename this function?

BenWilson2 replied:
much more clear decorator name. changed

harupy replied:
@BenWilson2 It appears it's not changed yet :)

harupy commented on examples/mlflow_artifacts/README.md:
````diff
@@ -16,16 +16,19 @@ First, launch the tracking server with the artifacts service via `mlflow server`
 ```sh
 # Launch a tracking server with the artifacts service
 $ mlflow server \
+    --serve-artifacts
````
```suggestion
    --serve-artifacts \
```

nit

BenWilson2 replied:
fixed

harupy commented on examples/mlflow_artifacts/docker-compose.yml:
```diff
@@ -54,6 +54,7 @@ services:
       --port 5500
       --artifacts-destination s3://bucket
       --gunicorn-opts "--log-level debug"
+      --serve-artifacts
```
```suggestion
      --serve-artifacts
      --artifacts-only
```

Can we also add `--artifacts-only`?

BenWilson2 replied:
added :)

BenWilson2 replied:
added

harupy commented on mlflow/server/handlers.py:
```diff
@@ -262,6 +262,43 @@ def wrapper(*args, **kwargs):
+
+        if not os.environ.get(SERVE_ARTIFACTS_ENV_VAR):
+            return Response(
+                (
+                    "Endpoints for mlflow artifacts service are disabled. To enable them, "
+                    "run `mlflow server` with `--serve-artfiacts-opt`"
+                ),
```
```suggestion
                (
                    "Endpoint disabled due to the mlflow server running without `--serve-artifacts`"
                    "To enable artifacts server functionality, run `mlflow server` with "
                    "`--serve-artifacts`"
                ),
```


Can we use a similar message as `_disable_mlflow_artifacts_only`?

harupy replied:
When we make an artifact request against mlflow server launched without `--serve-artifacts`, the error message looks like this:

```
raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 503 Server Error: SERVICE UNAVAILABLE for URL: 
http://localhost:5000/api/2.0/mlflow-artifacts/artifacts/experiments/0/15d993c8025f471a833dd97969920970/artifacts/a.txt
```

It doesn't explain why it's unavailable This is because we use `raise_for_status` to raise a request error:

https://github.com/mlflow/mlflow/blob/f6a6b20c1330a9c4808ad20ba88ec311a53fdc25/mlflow/store/artifact/http_artifact_repo.py#L28

We might need to use the approach in this StackOverflow answer: https://stackoverflow.com/a/52392219

BenWilson2 committed:
Merge branch 'master' of https://github.com/mlflow/mlflow into support-serve-artifacts

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
PR changes

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
PR updates

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

harupy commented on mlflow/cli.py:
```diff
@@ -279,7 +279,15 @@ def ui(backend_store_uri, default_artifact_root, artifacts_destination, port, ho
-            backend_store_uri, default_artifact_root, artifacts_destination, host, port, None, 1
+            backend_store_uri,
+            default_artifact_root,
+            "false",
+            "false",
```
```suggestion
            False,
            False,
```

Should these be `False`?

BenWilson2 replied:
It definitely should. I forgot about how Python will return a bool validation on a string of "true" or "false" the same as the bool operator.

harupy commented on mlflow/cli.py:
```diff
@@ -320,6 +328,25 @@ def _validate_static_prefix(ctx, param, value):  # pylint: disable=unused-argume
+    "by routing these requests to the storage location that is specified by "
+    "'--artifact-destination' directly through a proxy. The default location that "
+    "these requests are served from is a local './mlartifacts' directory which can be "
+    "overridden via '--artifact-destination' arguments. "
```
```suggestion
    "overridden via '--artifacts-destination' argument. "
```

nit

BenWilson2 replied:
fixed

harupy commented on tests/test_cli.py:
```diff
@@ -235,3 +249,46 @@ def predict(self, context, model_input):  # pylint: disable=unused-variable
+    port = get_safe_port()
+    cmd = ["mlflow", "server", "--port", str(port), "--artifacts-only"]
+    process = subprocess.Popen(cmd)
+    with pytest.raises(
+        HTTPError,
+        match="Endpoint disabled due to the mlflow server running in `--artifacts-only` mode.",
+    ):
+        _await_server_up_or_die(port, timeout=10)
+        resp = requests.get(f"http://localhost:{port}/api/2.0/mlflow/experiments/list")
+        augmented_raise_for_status(resp)
```
```suggestion
    _await_server_up_or_die(port, timeout=10)
    with pytest.raises(
        HTTPError,
        match="Endpoint disabled due to the mlflow server running in `--artifacts-only` mode.",
    ):
        resp = requests.get(f"http://localhost:{port}/api/2.0/mlflow/experiments/list")
        augmented_raise_for_status(resp)
```

Can we refactor the code here to clarify which line throws an `HTTPError`?

harupy replied:
```suggestion
    _await_server_up_or_die(port, timeout=10)
    resp = requests.get(f"http://localhost:{port}/api/2.0/mlflow/experiments/list")
    assert resp.status_code == 503
    assert "Endpoint disabled due to the mlflow server running in" in resp.text
```

On second thought, can we just check `resp.text` contains `Endpoint disabled due to the mlflow server running in`?

harupy replied:
```suggestion
    _await_server_up_or_die(port, timeout=10)
    resp = requests.get(f"http://localhost:{port}/api/2.0/mlflow/experiments/list")
    assert resp.status_code == 503
    assert "Endpoint disabled due to the mlflow server running in `--artifacts-only" in resp.text
```

On second thought, can we just check `resp.text` contains `Endpoint disabled due to the mlflow server running in`?

BenWilson2 replied:
changed that test and the test `test_mlflow_artifact_service_unavailable_without_config()`

harupy commented on examples/mlflow_artifacts/docker-compose.yml:
```diff
@@ -82,6 +84,7 @@ services:
       --backend-store-uri postgresql://user:password@postgres:5432/db
       --default-artifact-root http://artifacts-server:5500/api/2.0/mlflow-artifacts/artifacts/experiments
       --gunicorn-opts "--log-level debug"
+      --serve-artifacts
```
```suggestion
```

I think this can be removed because the client doesn't send requests for artifact operations to `tracking-server`.

harupy replied:
I locally tested and it worked without `--serve-artifacts`.

BenWilson2 replied:
removed

harupy commented on examples/mlflow_artifacts/example.py:
```diff
@@ -10,6 +10,10 @@ def save_text(path, text):
         f.write(text)
 
 
+#  NOTE: ensure the tracking server has been started with --serve-artifacts to enable
+#        MLflow artifact serving functionality.
```
Nice note!

harupy commented on tests/test_cli.py:
```diff
@@ -235,3 +249,46 @@ def predict(self, context, model_input):  # pylint: disable=unused-variable
+    cmd = ["mlflow", "server", "--port", str(port)]
+    process = subprocess.Popen(cmd)
+    try:
+        _await_server_up_or_die(port, timeout=10)
+        resp = requests.get(f"http://localhost:{port}/api/2.0/mlflow-artifacts/artifacts")
+        assert resp.status_code == 503
```
Can we check the error message as we do in `test_mlflow_tracking_disabled_in_artifacts_only_mode`?

BenWilson2 replied:
updated

harupy commented on mlflow/utils/rest_utils.py:
```diff
@@ -186,6 +187,17 @@ def verify_rest_response(response, endpoint):
+        response.raise_for_status()
+    except HTTPError as e:
+        if response.text:
+            raise HTTPError(f"{response.status_code} Error: {response.text}")
```
```suggestion
            raise HTTPError(f"{e}. Response text: {e.response.text}")
```

Can we include the original error message?

```
# Now
requests.exceptions.HTTPError: 503 Error: Endpoint disabled due to the mlflow server running without `--serve-artifacts`. To enable artifacts server functionaltiy, run `mlflow server` with `--serve-artfiacts`
```

```
# Proposed
requests.exceptions.HTTPError: 503 Server Error: SERVICE UNAVAILABLE for url: http://localhost:5000/api/2.0/mlflow-artifacts/artifacts/experiments/0/e6fdde74db494268812a4286cff5100d/artifacts/a.txt. Response text: Endpoint disabled due to the mlflow server running without `--serve-artifacts`. To enable artifacts server functionaltiy, run `mlflow server` with `--serve-artfiacts`
```

BenWilson2 replied:
much better to have the server uri and the full statement in there. Updated!

harupy commented on mlflow/server/handlers.py:
```diff
@@ -262,6 +262,44 @@ def wrapper(*args, **kwargs):
+        if not os.environ.get(SERVE_ARTIFACTS_ENV_VAR):
+            return Response(
+                (
+                    "Endpoint disabled due to the mlflow server running without "
```
```suggestion
                    f"Endpoint: {request.url_rule} disabled due to the mlflow server running without "
```

Can we include `request.url_rule` to give a more detailed error message?

BenWilson2 replied:
definitely. updated tests as well

BenWilson2 committed:
PR feedback changes

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
lint

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

harupy commented on mlflow/cli.py:
```diff
@@ -279,7 +279,15 @@ def ui(backend_store_uri, default_artifact_root, artifacts_destination, port, ho
-            backend_store_uri, default_artifact_root, artifacts_destination, host, port, None, 1
+            backend_store_uri,
+            default_artifact_root,
+            False,
```
Can we support `--serve-artifacts` for the `mlflow ui` command?

harupy commented on mlflow/server/handlers.py:
```diff
@@ -262,6 +262,44 @@ def wrapper(*args, **kwargs):
+        if os.environ.get(ARTIFACTS_ONLY_ENV_VAR):
+            return Response(
+                (
+                    "Endpoint disabled due to the mlflow server running in `--artifacts-only` "
```
```suggestion
                    f"Endpoint {request.url_rule} disabled due to the mlflow server running in `--artifacts-only` "
```

Can we also include `request.url_rule` here?

harupy commented:
LGTM, once the remaining comments are addressed! Thanks for working on this :)

@dbczumar Could you also take a look if you have a chance?

BenWilson2 committed:
Add ui server support for proxied artifacts and update exception messages

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

dbczumar reviewed (changes requested):
@BenWilson2 This is awesome! One point of feedback I noticed while manual QAing is that, when `--serve-artifacts` is specified, `--default-artifact-root` still defaults to `./mlruns`. For artifact serving, it should default to an `mlflow-artifacts:/` URI. Is support for `mlflow-artifacts` URIs coming in a separate PR? If so, let's merge that one first.

Once we change the default behavior, can we update the documentation for the `--default-artifact-root` option? Even for the existing behavior with file stores, the `mlflow server --help` output is pretty confusing and has some English syntax errors:

```
  --default-artifact-root URI  Local or S3 URI to store
                               artifacts, for new experiments.
                               Note that this flag does not
                               impact already-created
                               experiments. Default: Within
                               file store, if a file:/ URI is
                               provided. If a sql backend is
                               used, then this option is
                               required.
```

BenWilson2 committed:
Rebase and discard formatting commits

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

harupy commented on mlflow/cli.py:
```diff
@@ -263,11 +270,9 @@ def ui(backend_store_uri, default_artifact_root, artifacts_destination, port, ho
-        else:
-            default_artifact_root = DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH
+    default_artifact_root = resolve_default_artifact_root(
+        serve_artifacts, default_artifact_root, backend_store_uri, True
```
```suggestion
        serve_artifacts, default_artifact_root, backend_store_uri, resolve_to_local=True
```

Can we use a keyword argument here?

harupy commented on mlflow/server/handlers.py:
```diff
@@ -262,6 +262,44 @@ def wrapper(*args, **kwargs):
+                (
+                    f"Endpoint: {request.url_rule} disabled due to the mlflow server running "
+                    "without `--serve-artifacts`. To enable artifacts server functionaltiy, "
+                    "run `mlflow server` with `--serve-artfiacts`"
```
```suggestion
                    "run `mlflow server` with `--serve-artifacts`"
```

typo

harupy commented on mlflow/server/handlers.py:
```diff
@@ -262,6 +262,44 @@ def wrapper(*args, **kwargs):
+            return Response(
+                (
+                    f"Endpoint: {request.url_rule} disabled due to the mlflow server running "
+                    "without `--serve-artifacts`. To enable artifacts server functionaltiy, "
```
```suggestion
                    "without `--serve-artifacts`. To enable artifacts server functionality, "
```

typo

harupy commented:
@BenWilson2 Can we merge the master branch to include changes made by https://github.com/mlflow/mlflow/pull/5070?

harupy commented:
https://github.com/mlflow/mlflow/blob/776831a35c6460581c5b6642795926ebcb8b8624/mlflow/store/artifact/mlflow_artifacts_repo.py#L75

BenWilson2 committed:
typos

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'master' of https://github.com/mlflow/mlflow into support-serve-artifacts

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 commented:
> https://github.com/mlflow/mlflow/blob/776831a35c6460581c5b6642795926ebcb8b8624/mlflow/store/artifact/mlflow_artifacts_repo.py#L75
> 
> Is this line missing a slash after `{track_parse.path}`?
> 
> ```python
> resolved = f"{base_url}{track_parse.path}/{uri_parse.path.lstrip('/')}" 
> ```

the parsed `path` element will have a trailing '/' when returned from `urllib.parse.urlparse`. If we include another '/', we'll have paths like this:
Expected :http://myhostname/api/2.0/mlflow-artifacts/artifacts/my/artifact/path/host
**_Actual   :http://myhostname/api/2.0/mlflow-artifacts/artifacts//my/artifact/path/host_**

harupy commented:
@BenWilson2 My suggested code was incorrect. I think we need a slash before `track_parse.path` in a case where `tracking_uri` looks like `http://localhost:5000`.

```python
resolved = f"{base_url}/{track_parse.path}{uri_parse.path.lstrip('/')}" 
```

BenWilson2 committed:
fix parsing of uri for trailing slash

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 commented:
> @BenWilson2 My suggested code was incorrect. I think we need a slash before `track_parse.path` in a case where `tracking_uri` looks like `http://localhost:5000`.
> 
> ```python
> resolved = f"{base_url}/{track_parse.path}{uri_parse.path.lstrip('/')}" 
> ```
> 
> ```
> Expected: http://myhostname/api/2.0/mlflow-artifacts/artifacts/my/artifact/path/host
> Actual  : http://myhostname/api/2.0/mlflow-artifacts/artifactsmy/artifact/path/host
>                                                      ^^^^^^^^^^^
> ```

Ah, I get it! Added fixes for this and added fixture params to validate in the test suite.

harupy commented on tests/store/artifact/test_mlflow_artifact_repo.py:
```diff
@@ -8,11 +8,13 @@
 
-@pytest.fixture(scope="module", autouse=True)
-def set_tracking_uri():
+@pytest.fixture(
+    scope="module", autouse=True, params=["http://localhost:5000", "http://localhost:5000/"]
+)
```
This doubles the number of tests. Can we just test that both `http://localhost:5000` and `http://localhost:5000/` can be resolved correctly?

BenWilson2 replied:
definitely!

BenWilson2 replied:
Created another fixture and used an explicit single test for validating alternate uri naming convention

BenWilson2 committed:
Test complexity reduction

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
tracking uri validation checks and test simplification

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

harupy commented on tests/store/artifact/test_mlflow_artifact_repo.py:
```diff
@@ -29,29 +29,29 @@ def test_mlflow_artifact_uri_formats_resolved():
         (
             f"mlflow-artifacts://myhostname:4242{base_path}/hostport",
             f"http://myhostname:4242{base_url}{base_path}/hostport",
+            "http://myhostname:4242",
```
Can we parametrize this test?

BenWilson2 replied:
definitely. Added.

BenWilson2 committed:
Parameterize the uri resolution tests to aid in test debugging

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

harupy commented on tests/store/artifact/test_mlflow_artifact_repo.py:
```diff
@@ -22,36 +22,43 @@ def test_artifact_uri_factory():
+base_path = "/my/artifact/path"
+conditions = [
+    (
+        f"mlflow-artifacts://myhostname:4242{base_path}/hostport",
+        f"http://myhostname:4242{base_url}{base_path}/hostport",
+        "http://myhostname:4242",
```
```suggestion
        f"mlflow-artifacts://myhostname:4242{base_path}/hostport",
        "http://myhostname:4242",
        f"http://myhostname:4242{base_url}{base_path}/hostport",
```

Can we fix the order?

BenWilson2 replied:
I like your edited way in the last comment better. Changed to that easier to read and cleaner implementation.

BenWilson2 committed:
Cleaner test syntax

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

harupy commented:
LGTM!

dbczumar commented on mlflow/store/tracking/__init__.py:
```diff
@@ -10,5 +10,6 @@
 # Also used as default location for artifacts, when not provided, in non local file based backends
 # (eg MySQL)
 DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH = "./mlruns"
+DEFAULT_ARTIFACTS_URI = "mlflow-artifacts:/"
```
Can we add an inline comment explaining what this is used for?

BenWilson2 replied:
certainly!

dbczumar commented on mlflow/cli.py:
```diff
@@ -233,20 +233,27 @@ def _validate_server_args(gunicorn_opts=None, workers=None, waitress_opts=None):
+    "server backends that rely on SQL, this option is required in order to store artifacts. "
+    "Note that this flag does not impact already-created experiments with any previous "
+    "configuration of an MLflow server instance. "
+    f"By default, data will be logged to the {DEFAULT_ARTIFACTS_URI} uri proxy if "
+    "the --serve-artifacts option is enabled. Otherwise, the default location will "
+    f"be {DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH}.",
```
```suggestion
    f"If the --serve-artifacts option is specified, the default artifact root is {DEFAULT_ARTIFACTS_URI}. "
    f "otherwise, the default artifact root is {DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH}".
```

dbczumar reviewed (approved):
LGTM with 2 small docs comments. Thanks so much, @BenWilson2 !

BenWilson2 committed:
Add and adjust pydoc strings

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>
